### DISCUSSION TITLE
@swc/core vs. swc_core

### DISCUSSION THREAD

**Discussion by LekoArts** [swc-project/swc#5638] [Q&A] [answered]
Hi!

**TL;DR:** Is using `@swc/core` instead of `swc_core` (and Rust toolchain) okay as a trade-off to have faster migration?

I want to replace some custom Babel plugins (that transform their input, e.g. remove imports and their definitions) that are used in an npm package by a variety of users. While I know some Rust I'm by no means an expert and other possible contributors to that code also don't really know Rust.

I couldn't get an answer from the docs for when to use https://swc.rs/docs/usage/core and when to write a transformer like shown in https://swc.rs/docs/plugin/ecmascript/getting-started.

While I'd like to get better at Rust and choosing the second option will surely help with that, the first option seems like a safer bet as:

- Looking at an example like https://github.com/Sociosarbis/swc-plugin-import/blob/master/src/index.ts it seems the code is really similar to existing Babel plugins
- No Rust toolchain, no build step for bindings, no handling needed for wasm (?)

Are there any downsides to using `@swc/core` instead of writing it all in Rust? I guess it's performance? How much of a performance gap would it be?

Would https://napi.rs/ help with distributing the plugin to more platforms and better automatic handling of the WASM binary loading?

All in all I'm just looking for the potential trade-offs and if they are worth it, because if the gap isn't that big then the added benefit of being able to migrate faster from Babel to SWC is worth it.

Thanks!

**Comment by kdy1** [comment, marked as answer]
JS plugins will be dropped once we finish other works for v2

**Comment by kwonoj** [comment]
I'm not sure what 

> @swc/core instead of swc_core 

exactly means.

`swc_core` is a package to build binary (`@swc/core`, or plugins, or custom binaries) and they are not 1:1 exchangeable in most of the context. If your intention is having some custom transform / or plugin, `swc_core` is the right package you are looking for - `@swc/core`, node.js binding will consume those plugins.

**Reply by LekoArts** [reply]
Okay, thanks for that explanation, that makes more sense. I guess I was mostly confused by https://github.com/Sociosarbis/swc-plugin-import/blob/master/src/index.ts - so that will be removed in v2, right? This is where my assumption came from that for the transform use case they are "kinda" equivalent.

And the flow should be:

1. Create transformer/plugin with `swc_core`
2. `@swc/core` consumes this plugin

Is there an example/doc on how to combine those two together? In https://swc.rs/docs/usage/core#options I see `jsc?: JscConfig;` so I suppose I could pass in something similar then like shown in e.g. https://github.com/swc-project/plugins/tree/main/packages/styled-components (I don't want to use a `.swcrc` file)

**Reply by kwonoj** [reply]
There is jsc.experimental.plugins config options can be passed into napi interfaces in `@swc/core`'s transform.

**Reply by LekoArts** [reply]
Okay, thank you 👍
